template/template.go

```diff
@@ -42,16 +42,29 @@ type Template struct {
 	ExternalURL *url.URL
 }
 
+// FromGlobsWithAdditionalFuncs calls ParseGlob on all path globs provided and returns the
+// resulting Template with custom functions added, DefaultFuncs have precedence over the additional functions.
+func FromGlobsWithAdditionalFuncs(funcs FuncMap, paths ...string) (*Template, error) {
+	for n, f := range DefaultFuncs {
+		funcs[n] = f
+	}
+	return initTemplateFromGlob(paths, funcs)
+}
+
 // FromGlobs calls ParseGlob on all path globs provided and returns the
 // resulting Template.
 func FromGlobs(paths ...string) (*Template, error) {
```

Comment:
I'm not fond of the idea of having unused functions in the code - it's hard to know who's using them or not and whether they serve their original purpose. 

The surface area of `FromGlobs` is not _that_ big, I think I would instead break the API now in favour of something that's a bit more open/close such as `FromGlob(paths []string, options ...Option)`.

As such, my suggestion for this would be:

```golang
type Option func(template *Template)

func WithFunctions(funcs FuncMap) Option {
	return func(template *Template) {
		for n, f := range DefaultFuncs {
			funcs[n] = f
		}

		template.text = template.text.Funcs(tmpltext.FuncMap(funcs))
		template.html = template.html.Funcs(tmpltext.FuncMap(funcs))
	}
}

// FromGlobs calls ParseGlob on all path globs provided and returns the
// resulting Template.
func FromGlobs(paths []string, options ...Option) (*Template, error) {
	t := &Template{
		text: tmpltext.New("").Option("missingkey=zero"),
		html: tmplhtml.New("").Option("missingkey=zero"),
	}
	for _, o := range options {
		o(t)
	}

```

Reply:
Thanks for the review!
Understood, should be refactored to use options now PTAL

Reply:
:thinking: At first I actually added the `WithFunctions` option implementation also to the Alertmanger code base, but I suspect that would be another unused function, so I dropped it here and will be implemented on the side who needs it.

Reply:
Sadly, given both `text` and `html` are private you won't be able to create an option that modifies both attributes of the struct.

My suggestion is something like this:

```
// Option is generic modifier of the Template
type Option func(template *Template)

var DefaultOptions = []Option{withFunctions(DefaultFuncs)}

func withFunctions(funcs FuncMap) Option {
	return func(template *Template) {
		template.text = template.text.Funcs(tmpltext.FuncMap(funcs))
		template.html = template.html.Funcs(tmplhtml.FuncMap(funcs))
	}
}

// FromGlobs calls ParseGlob on all path globs provided and returns the
// resulting Template. Additional Options can be passed to it to extend the default Template.
func FromGlobs(paths []string, options ...Option) (*Template, error) {
	t := &Template{
		text: tmpltext.New("").Option("missingkey=zero"),
		html: tmplhtml.New("").Option("missingkey=zero"),
	}

	if len(options) == 0 {
		options = DefaultOptions
	}

	for _, o := range options {
		o(t)
	}
```

Reply:
Actually, not sure if using the `Option` passing it the Template with almost all unexported attributes is useful at all. 
The last suggestion, if I'm not mistaken, would more or less just justified the presence of the `WithFunctions` option in the alertmanager codebase by using it for the default?

Anyway, refactored as suggested (the `withFunctions` is exported since otherwise I still do not see hot to use it outside this codebase)

But maybe the option could rather be  `Option func(text *tmpltext.Template, html *tmplhtml.Template)` making it possible to implement it outside Alertmanager codebase and generally more useful, or stick to the originally suggested  `FromGlob(paths []string, additionalFunctions FuncMap)` and do not overcomplicate things? :shrug:

Reply:
Ok, so refactored hopefully for the last time :grin: PTAL